src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -371,4 +405,86 @@ isMissingPackets = do
       Nothing -> pure True
       Just ct -> do
         len <- lengthPacketSet pxs
-        pure (fromIntegral len - 1 < ct)
\ No newline at end of file
+        pure (fromIntegral len - 1 < ct)
+
+-- Unsafe ----------------------------------------------------------------------
+
+fromPacketBuildR ::
+  MonadIO m =>
+  Ptr Word8 ->
+  Int ->
+  Packet ByteString ->
+  m (Ptr Word8, Int)
+fromPacketBuildR buffer unused packet = do
+  let build = Encode.reverseMessageBuilder (wireBuildPacket packet)
+  liftIO (fromBuildR build buffer unused)
+
+encodeBufferOffPacket ::
+  MonadIO m =>
+  Ptr Word8 ->
+  Int ->
+  Packet ByteString ->
+  m ByteString
+encodeBufferOffPacket ptr unused packet = do
+  (ptr', unused') <- fromPacketBuildR ptr unused packet
+  size <- liftIO (readTotal ptr' unused')
+  liftIO (unsafePackCStringLen ptr' size)
```

Comment:
It looks like this code is assuming that there is no reallocation.  If reallocation occurs then there will be a problem, because `sealBuffer#` will assume that it can just take control of the `MutableByteArray` and convert it into a `ByteString` chunk, without any danger of some other code reusing it.  Furthermore, it will start writing to a newly-allocated buffer.

Reply:
Thats correct, no reallocation occurs.

Reply:
OK, then I guess it is OK, but perhaps there should be some comments laying out why it is guaranteed that the encoded size will be smaller than the initial buffer size?

Reply:
Sure, will do.